numba_dpex/experimental/launcher.py

```diff
@@ -0,0 +1,295 @@
+
+from numba_dpex import config, dpjit
+from numba_dpex.core.exceptions import UnreachableError
+from numba_dpex.core.targets.kernel_target import DpexKernelTargetContext
+from numba_dpex.core.types import DpnpNdArray, NdRangeType, RangeType
+from numba_dpex.core.utils import kernel_launcher as kl
+from numba_dpex.dpctl_iface import libsyclinterface_bindings as sycl
+from numba_dpex.experimental.kernel_dispatcher import _KernelModule
+from numba_dpex.utils import create_null_ptr
+
+
+def _get_queue_ref_val(
+    kernel_targetctx: DpexKernelTargetContext,
+    builder: llvmir.IRBuilder,
+    kernel_argtys: [types.Type, ...],
+    kernel_args,
+):
+    """
+    Get the sycl queue from the first DpnpNdArray argument. Prior passes
+    before lowering make sure that compute-follows-data is enforceable
+    for a specific call to a kernel. As such, at the stage of lowering
+    the queue from the first DpnpNdArray argument can be extracted.
+    """
+
+    for arg_num, argty in enumerate(kernel_argtys):
+        if isinstance(argty, DpnpNdArray):
+            llvm_val = kernel_args[arg_num]
+            datamodel = kernel_targetctx.data_model_manager.lookup(argty)
+            sycl_queue_attr_pos = datamodel.get_field_position("sycl_queue")
+            ptr_to_queue_ref = builder.extract_value(
+                llvm_val, sycl_queue_attr_pos
+            )
+            break
+
+    return ptr_to_queue_ref
+
+
+def _create_kernel_bundle_from_spirv(
+    builder: llvmir.IRBuilder,
+    targetctx: cpu.CPUContext,
+    queue_ref: llvmir.PointerType,
+    kernel_bc: llvmir.Constant,
+    kernel_bc_size_in_bytes: int,
+):
+    dref = sycl.dpctl_queue_get_device(builder, queue_ref)
+    cref = sycl.dpctl_queue_get_context(builder, queue_ref)
+    args = [
+        cref,
+        dref,
+        kernel_bc,
+        llvmir.Constant(llvmir.IntType(64), kernel_bc_size_in_bytes),
+        builder.load(create_null_ptr(builder, targetctx)),
+    ]
+    kbref = sycl.dpctl_kernel_bundle_create_from_spirv(builder, *args)
+    sycl.dpctl_context_delete(builder, cref)
+    sycl.dpctl_device_delete(builder, dref)
+
+    return kbref
+
+
+def _get_num_flattened_kernel_args(
+    kernel_targetctx: DpexKernelTargetContext,
+    kernel_argtys: tuple[types.Type, ...],
+):
+    num_flattened_kernel_args = 0
+    for arg_type in kernel_argtys:
+        if isinstance(arg_type, DpnpNdArray):
+            datamodel = kernel_targetctx.data_model_manager.lookup(arg_type)
+            num_flattened_kernel_args += datamodel.flattened_field_count
+        elif arg_type == types.complex64 or arg_type == types.complex128:
+            num_flattened_kernel_args += 2
+        else:
+            num_flattened_kernel_args += 1
+
+    return num_flattened_kernel_args
+
+
+def _create_kernel_launcher_body(
+    codegen_targetctx: cpu.CPUContext,
+    kernel_targetctx: DpexKernelTargetContext,
+    builder: llvmir.IRBuilder,
+    indexer_argty: RangeType | NdRangeType,
```

Comment:
```suggestion
    indexer_argty: Union[RangeType, NdRangeType],
```
`Union` is available from `typing`:
```
from typing import Union
``
https://peps.python.org/pep-0604/